store/src/java/com/zimbra/cs/mailbox/cache/MapItemCache.java

```diff
@@ -0,0 +1,83 @@
+package com.zimbra.cs.mailbox.cache;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import com.zimbra.cs.mailbox.MailItem;
+import com.zimbra.cs.mailbox.Mailbox;
+
+public abstract class MapItemCache<T> extends ItemCache {
+    private final Map<Integer /* id */, T> mapById;
+    private final Map<String /* uuid */, Integer /* id */> uuid2id;
+
+    public MapItemCache(Mailbox mbox, Map<Integer, T> itemMap, Map<String, Integer> uuidMap) {
+        super(mbox);
+        mapById = itemMap;
+        uuid2id = uuidMap;
+    }
+
+    protected abstract T toCacheValue(MailItem item);
+
+    protected abstract MailItem fromCacheValue(T value);
+
+    @Override
+    public void put(MailItem item) {
+        int id = item.getId();
+        mapById.put(id, toCacheValue(item));
+        String uuid = item.getUuid();
+        if (uuid != null) {
+            uuid2id.put(uuid, id);
```

Comment:
Should an exception be raised If the item has no uuid and can't be stored?

Reply:
I don't think so - this is actually exactly now the original `ItemCache` worked, minus the `toCacheValue` call.